processor/probabilisticsamplerprocessor/sampler_mode.go

```diff
@@ -0,0 +1,316 @@
+	return th.tvalueThreshold
+}
+
+// neverSampler always decides false.
+type neverSampler struct {
+}
+
+func (*neverSampler) decide(_ samplingCarrier) sampling.Threshold {
+	return sampling.NeverSampleThreshold
+}
+
+func getBytesFromValue(value pcommon.Value) []byte {
+	if value.Type() == pcommon.ValueTypeBytes {
+		return value.Bytes().AsRaw()
+	}
+	return []byte(value.AsString())
+}
+
+func randomnessFromBytes(b []byte, hashSeed uint32) sampling.Randomness {
+	hashed32 := computeHash(b, hashSeed)
+	hashed := uint64(hashed32 & bitMaskHashBuckets)
+
+	// Ordinarily, hashed is compared against an acceptance
+	// threshold i.e., sampled when hashed < scaledSamplerate,
+	// which has the form R < T with T in [1, 2^14] and
+	// R in [0, 2^14-1].
+	//
+	// Here, modify R to R' and T to T', so that the sampling
+	// equation has identical form to the specification, i.e., T'
+	// <= R', using:
+	//
+	//   T' = numHashBuckets-T
+	//   R' = numHashBuckets-1-R
+	//
+	// As a result, R' has the correct most-significant 14 bits to
+	// use in an R-value.
+	rprime14 := numHashBuckets - 1 - hashed
+
+	// There are 18 unused bits from the FNV hash function.
+	unused18 := uint64(hashed32 >> (32 - numHashBucketsLg2))
+	mixed28 := unused18 ^ (unused18 << 10)
+
+	// The 56 bit quantity here consists of, most- to least-significant:
+	// - 14 bits: R' = numHashBuckets - 1 - hashed
+	// - 28 bits: mixture of unused 18 bits
+	// - 14 bits: original `hashed`.
+	rnd56 := (rprime14 << 42) | (mixed28 << 14) | hashed
+
+	// Note: by construction:
+	// - OTel samplers make the same probabilistic decision with this r-value,
+	// - only 14 out of 56 bits are used in the sampling decision,
+	// - there are only 32 actual random bits.
+	rnd, _ := sampling.UnsignedToRandomness(rnd56)
+	return rnd
+}
+
+// consistencyCheck checks for certain inconsistent inputs.
+//
+// if the randomness is missing, returns ErrMissingRandomness.
+func consistencyCheck(rnd randomnessNamer, _ samplingCarrier) error {
+	if isMissing(rnd) {
+		return ErrMissingRandomness
+	}
+	return nil
+}
+
+// makeSample constructs a sampler. There are no errors, as the only
+// potential error, out-of-range probability, is corrected automatically
+// according to the README, which allows percents >100 to equal 100%.
+//
+// Extending this logic, we round very small probabilities up to the
+// minimum supported value(s) which varies according to sampler mode.
+func makeSampler(cfg *Config) dataSampler {
+	// README allows percents >100 to equal 100%.
+	pct := cfg.SamplingPercentage
+	if pct > 100 {
+		pct = 100
+	}
+
+	never := &neverSampler{}
+
+	if pct == 0 {
+		return never
+	}
+
+	// Note: the original hash function used in this code
+	// is preserved to ensure consistency across updates.
+	//
+	//   uint32(pct * percentageScaleFactor)
+	//
+	// (a) carried out the multiplication in 32-bit precision
+	// (b) rounded to zero instead of nearest.
+	scaledSamplerate := uint32(pct * percentageScaleFactor)
```

Comment:
I feel bad for making this suggestion, but I can't help:

```suggestion
	scaledSampleRate := uint32(pct * percentageScaleFactor)
```

Reply:
:+1: